server/src/main/java/org/elasticsearch/cluster/coordination/ElectionScheduler.java

```diff
@@ -0,0 +1,246 @@
+        }
+    }
+
+    private static final class ElectionMaxRetryIntervalSettingValidator implements Setting.Validator<TimeValue> {
+        @Override
+        public void validate(final TimeValue value, final Map<Setting<TimeValue>, TimeValue> settings) {
+            validateSettings(settings.get(ELECTION_MIN_RETRY_INTERVAL_SETTING), value);
+        }
+
+        @Override
+        public Iterator<Setting<TimeValue>> settings() {
+            return Collections.singletonList(ELECTION_MIN_RETRY_INTERVAL_SETTING).iterator();
+        }
+    }
+
+    static void validateSettings(final TimeValue electionMinRetryInterval, final TimeValue electionMaxRetryInterval) {
+        if (electionMaxRetryInterval.millis() < electionMinRetryInterval.millis() + 100) {
+            throw new IllegalArgumentException(validationExceptionMessage(
+                electionMinRetryInterval.toString(), electionMaxRetryInterval.toString()));
+        }
+    }
+
+    static String validationExceptionMessage(final String electionMinRetryInterval, final String electionMaxRetryInterval) {
+        return new ParameterizedMessage(
+            "Invalid election retry intervals: [{}] is [{}] and [{}] is [{}], but [{}] should be at least 100ms longer than [{}]",
+            ELECTION_MIN_RETRY_INTERVAL_SETTING_KEY, electionMinRetryInterval,
+            ELECTION_MAX_RETRY_INTERVAL_SETTING_KEY, electionMaxRetryInterval,
+            ELECTION_MAX_RETRY_INTERVAL_SETTING_KEY, ELECTION_MIN_RETRY_INTERVAL_SETTING_KEY).getFormattedMessage();
+    }
+
+    private Object currentScheduler; // only care about its identity
+    private final TimeValue electionMinRetryInterval;
+    private final TimeValue electionMaxRetryInterval;
+    private final Object mutex = new Object();
+    private final ThreadPool threadPool;
+    private final Random random;
+
+    private long currentDelayMillis;
+
+    ElectionScheduler(Settings settings, Random random, ThreadPool threadPool) {
+        super(settings);
+
+        this.random = random;
+        this.threadPool = threadPool;
+
+        synchronized (mutex) {
+            electionMinRetryInterval = ELECTION_MIN_RETRY_INTERVAL_SETTING.get(settings);
+            electionMaxRetryInterval = ELECTION_MAX_RETRY_INTERVAL_SETTING.get(settings);
+        }
+    }
+
+    public void start() {
+        logger.trace("starting");
+        final Runnable newScheduler;
+        synchronized (mutex) {
+            currentDelayMillis = electionMinRetryInterval.millis();
+            assert currentScheduler == null;
+            currentScheduler = newScheduler = new Runnable() {
+                @Override
+                public String toString() {
+                    return "ElectionScheduler: next election scheduler";
+                }
+
+                private boolean isCurrentScheduler() {
+                    return this == currentScheduler;
+                }
+
+                private void scheduleNextElection() {
```

Comment:
too much indentation. Can you move this method to the ElectionScheduler top-level class, and give it a `BooleanSupplier isActive` parameter? The runnable will then simply be

```
currentScheduler = newScheduler = new Runnable() {
    @Override
    public String toString() {
        return "ElectionScheduler: next election scheduler";
    }

    @Override
    public void run() {
        scheduleNextElection(() -> this == currentScheduler);
    }
};
```

Reply:
Yes, that's nicer.